File: sdk/go/watch_source.go
package aghsdk

import (
	"context"
	"encoding/json"
	"slices"
	"strings"
)

const extensionKindKey = "kind"

type registeredWatchSource struct {
	kind    string
	handler func(context.Context, rawWatchSourceRequest) (WatchPollResponse, error)
}

type rawWatchSourceRequest struct {
	kind                string
	spec                json.RawMessage
	expectedStateDigest string
	context             ExtensionContext
}

// WatchSource registers a raw JSON watch-source handler on the extension.
func (e *Extension) WatchSource(
	kind string,
	options WatchSourceOptions,
	fn WatchSourceHandlerFunc[json.RawMessage],
) error {
	return WatchSource[json.RawMessage](e, kind, options, fn)
}

// WatchSource registers a typed Go function as a loop watch-source handler.
func WatchSource[TSpec any](
	extension *Extension,
	kind string,
	options WatchSourceOptions,
	fn WatchSourceHandlerFunc[TSpec],
) error {
	if extension == nil {
		return NewInternalError("extension is required")
	}
	if fn == nil {
		return NewInvalidParamsError("watch source handler function is required", nil)
	}
	return extension.registerWatchSource(kind, options, func(
		ctx context.Context,
		req rawWatchSourceRequest,
	) (WatchPollResponse, error) {
		var spec TSpec
		rawSpec := req.spec
		if len(rawSpec) == 0 {
			rawSpec = json.RawMessage(`{}`)
		}
		if err := json.Unmarshal(rawSpec, &spec); err != nil {
			return WatchPollResponse{}, NewInvalidParamsError(
				"watch source spec does not match handler type",
				map[string]any{
					extensionKindKey:  req.kind,
					extensionErrorKey: err.Error(),
				},
			)
		}
		return fn(ctx, WatchSourceRequest[TSpec]{
			Kind:                req.kind,
			Spec:                spec,
			RawSpec:             cloneRawMessage(rawSpec),
			ExpectedStateDigest: req.expectedStateDigest,
			Context:             req.context,
			Host:                req.context.Host,
			Session:             req.context.Session,
		})
	})
}

// GetWatchSourceKinds returns the sorted watch-source kinds registered by WatchSource.
func (e *Extension) GetWatchSourceKinds() []string {
	if e == nil {
		return nil
	}
	e.mu.RLock()
	defer e.mu.RUnlock()
	return e.watchSourceKindsLocked()
}

func (e *Extension) registerWatchSource(
	kind string,
	_ WatchSourceOptions,
	fn func(context.Context, rawWatchSourceRequest) (WatchPollResponse, error),
) error {
	cleanKind := strings.TrimSpace(kind)
	if cleanKind == "" {
		return NewInvalidParamsError("watch source kind is required", nil)
	}
	if fn == nil {
		return NewInvalidParamsError("watch source handler function is required", nil)
	}

	e.mu.Lock()
	defer e.mu.Unlock()
	if e.initialized {
		return NewInvalidParamsError("extension registration is closed after initialize", nil)
	}
	if err := e.ensureWatchSourceRegistrationAvailableLocked(cleanKind); err != nil {
		return err
	}
	e.watchHandlers[cleanKind] = registeredWatchSource{kind: cleanKind, handler: fn}
	e.definition.Capabilities.Provides = normalizeStrings(
		append(e.definition.Capabilities.Provides, CapabilityProvideWatchSource),
	)
	e.transport.Handle(ExtensionServiceMethodWatchPoll, e.dispatch)
	return nil
}

func (e *Extension) ensureWatchSourceRegistrationAvailableLocked(kind string) error {
	if _, ok := e.watchHandlers[kind]; ok {
		return NewInvalidParamsError("watch source already registered", map[string]any{extensionKindKey: kind})
	}
	if _, ok := e.handlers[ExtensionServiceMethodWatchPoll]; ok {
		return NewInvalidParamsError("watch/poll is reserved by WatchSource", nil)
	}
	return nil
}

func (e *Extension) handleWatchPoll(
	ctx context.Context,
	request JSONRPCRequestEnvelope,
	params json.RawMessage,
) (WatchPollResponse, error) {
	var poll WatchPollRequest
	if err := json.Unmarshal(params, &poll); err != nil {
		return WatchPollResponse{}, NewInvalidParamsError(
			"watch/poll params must be an object",
			map[string]any{extensionErrorKey: err.Error()},
		)
	}
	kind, err := watchSourceKind(poll.Spec)
	if err != nil {
		return WatchPollResponse{}, err
	}
	e.mu.RLock()
	registered, ok := e.watchHandlers[kind]
	e.mu.RUnlock()
	if !ok {
		return WatchPollResponse{}, NewInvalidParamsError(
			"watch source kind is not registered",
			map[string]any{extensionKindKey: kind},
		)
	}
	contextValue := e.makeContext(request)
	return registered.handler(ctx, rawWatchSourceRequest{
		kind:                registered.kind,
		spec:                cloneRawMessage(poll.Spec),
		expectedStateDigest: strings.TrimSpace(poll.ExpectedStateDigest),
		context:             contextValue,
	})
}

func (e *Extension) watchSourceKindsLocked() []string {
	if len(e.watchHandlers) == 0 {
		return nil
	}
	out := make([]string, 0, len(e.watchHandlers))
	for kind := range e.watchHandlers {
		out = append(out, kind)
	}
	slices.Sort(out)
	return out
}

func isWatchSourceMethod(method string) bool {
	return method == ExtensionServiceMethodWatchPoll
}

func watchSourceKind(spec json.RawMessage) (string, error) {
	if len(spec) == 0 || !json.Valid(spec) {
		return "", NewInvalidParamsError("watch source spec must be valid JSON", nil)
	}
	var payload struct {
		Kind string `json:"kind"`
	}
	if err := json.Unmarshal(spec, &payload); err != nil {
		return "", NewInvalidParamsError(
			"watch source spec must be an object",
			map[string]any{extensionErrorKey: err.Error()},
		)
	}
	kind := strings.TrimSpace(payload.Kind)
	if kind == "" {
		return "", NewInvalidParamsError("watch source spec.kind is required", nil)
	}
	return kind, nil
}

File: sdk/typescript/src/watch-source.ts
import { InvalidParamsError } from "./errors.js";
import { cloneJSON } from "./json.js";
import type { HostAPI } from "./host-api.js";
import type { ExtensionContext, ExtensionSession } from "./extension.js";
import type { ExtensionDefinition, JSONRPCRequestEnvelope, JSONValue } from "./types.js";

export const WATCH_SOURCE_CAPABILITY = "loop.watch_source";
export const WATCH_POLL_METHOD = "watch/poll";

export interface ExtensionWatchSourceOptions {}

export interface WatchPollRequest {
  spec: JSONValue;
  expected_state_digest?: string;
}

export interface WatchPollResponse {
  ready: boolean;
  state_digest?: string;
  payload?: JSONValue;
  settled_at?: string;
}

export interface ExtensionWatchSourceContext<TSpec extends JSONValue = JSONValue> {
  readonly kind: string;
  readonly spec: TSpec;
  readonly expectedStateDigest?: string;
  readonly context: ExtensionContext;
  readonly host: HostAPI;
  readonly session: ExtensionSession;
}

export type ExtensionWatchSourceHandler<TSpec extends JSONValue = JSONValue> = (
  request: ExtensionWatchSourceContext<TSpec>
) => Promise<WatchPollResponse> | WatchPollResponse;

export interface RegisteredWatchSource {
  kind: string;
  handler: ExtensionWatchSourceHandler;
}

interface ExtensionWatchSourcesOptions {
  bindMethod: (method: string) => void;
  hasUserHandler: (method: string) => boolean;
  makeContext: (request: JSONRPCRequestEnvelope) => ExtensionContext;
}

export class ExtensionWatchSources {
  private readonly handlers = new Map<string, RegisteredWatchSource>();

  public constructor(private readonly options: ExtensionWatchSourcesOptions) {}

  public register<TSpec extends JSONValue = JSONValue>(
    definition: ExtensionDefinition,
    kind: string,
    _registerOptions: ExtensionWatchSourceOptions,
    handler: ExtensionWatchSourceHandler<TSpec>
  ): void {
    const cleanKind = kind.trim();
    if (!cleanKind) {
      throw new Error("watch source kind is required");
    }
    if (this.handlers.has(cleanKind)) {
      throw new Error(`watch source already registered: ${cleanKind}`);
    }
    if (this.options.hasUserHandler(WATCH_POLL_METHOD)) {
      throw new Error("watch/poll is reserved by extension.watchSource()");
    }
    this.handlers.set(cleanKind, {
      kind: cleanKind,
      handler: handler as ExtensionWatchSourceHandler,
    });
    this.ensureCapability(definition);
    this.options.bindMethod(WATCH_POLL_METHOD);
  }

  public hasHandlers(): boolean {
    return this.handlers.size > 0;
  }

  public implementedMethods(): string[] {
    return this.hasHandlers() ? [WATCH_POLL_METHOD] : [];
  }

  public bindMethods(): void {
    if (this.hasHandlers()) {
      this.options.bindMethod(WATCH_POLL_METHOD);
    }
  }

  public kinds(): string[] {
    return Array.from(this.handlers.keys()).sort();
  }

  public initializeResponseFields(): { watch_source_kinds?: string[] } {
    return this.hasHandlers() ? { watch_source_kinds: this.kinds() } : {};
  }

  public async handlePoll(
    request: JSONRPCRequestEnvelope,
    params: unknown
  ): Promise<WatchPollResponse> {
    const poll = parseWatchPollRequest(params);
    const registered = this.handlers.get(poll.kind);
    if (!registered) {
      throw new InvalidParamsError("watch source kind is not registered", { kind: poll.kind });
    }

    const context = this.options.makeContext(request);
    const response = await registered.handler({
      kind: registered.kind,
      spec: poll.spec,
      context,
      host: context.host,
      session: context.session,
      ...(poll.expected_state_digest ? { expectedStateDigest: poll.expected_state_digest } : {}),
    });
    return normalizeWatchPollResponse(response);
  }

  private ensureCapability(definition: ExtensionDefinition): void {
    const capabilities = definition.capabilities ?? {};
    capabilities.provides = normalizeStringList([
      ...(capabilities.provides ?? []),
      WATCH_SOURCE_CAPABILITY,
    ]);
    definition.capabilities = capabilities;
  }
}

export function isWatchSourceMethod(method: string): boolean {
  return method === WATCH_POLL_METHOD;
}

export function parseWatchPollRequest(params: unknown): WatchPollRequest & { kind: string } {
  if (typeof params !== "object" || params === null) {
    throw new InvalidParamsError("watch/poll params must be an object");
  }
  const request = params as WatchPollRequest;
  if (typeof request.spec !== "object" || request.spec === null || Array.isArray(request.spec)) {
    throw new InvalidParamsError("spec must be an object");
  }
  const kind = (request.spec as Record<string, JSONValue>).kind;
  if (typeof kind !== "string" || kind.trim() === "") {
    throw new InvalidParamsError("spec.kind is required");
  }
  if (
    request.expected_state_digest !== undefined &&
    typeof request.expected_state_digest !== "string"
  ) {
    throw new InvalidParamsError("expected_state_digest must be a string");
  }
  return {
    spec: cloneJSON(request.spec),
    kind: kind.trim(),
    ...(request.expected_state_digest
      ? { expected_state_digest: request.expected_state_digest.trim() }
      : {}),
  };
}

export function normalizeWatchPollResponse(response: WatchPollResponse): WatchPollResponse {
  if (typeof response !== "object" || response === null) {
    throw new InvalidParamsError("watch source response must be an object");
  }
  if (typeof response.ready !== "boolean") {
    throw new InvalidParamsError("watch source response.ready must be a boolean");
  }
  if (response.state_digest !== undefined && typeof response.state_digest !== "string") {
    throw new InvalidParamsError("watch source response.state_digest must be a string");
  }
  if (response.settled_at !== undefined && typeof response.settled_at !== "string") {
    throw new InvalidParamsError("watch source response.settled_at must be a string");
  }
  return {
    ready: response.ready,
    ...(response.state_digest ? { state_digest: response.state_digest } : {}),
    ...(response.payload !== undefined ? { payload: cloneJSON(response.payload) } : {}),
    ...(response.settled_at ? { settled_at: response.settled_at } : {}),
  };
}

function normalizeStringList(values: readonly string[] | undefined): string[] {
  return Array.from(new Set((values ?? []).map(value => value.trim()).filter(Boolean))).sort();
}
